feat: support .vue files

zcf0508 opened the pull request:
|default|prettify-ts|
|--|--|
|![image](https://github.com/user-attachments/assets/dfca3329-f285-4f61-aff5-1b1bc4112738)|![image](https://github.com/user-attachments/assets/67069cac-81eb-4277-b80b-406bdc47736c)|


close #55

zcf0508 committed:
feat: support .vue files

zcf0508 committed:
fix: postion err on windows

mylesmmurphy commented:
Hey @zcf0508! 😄 Thanks for opening this. At first glance it looks good, I like the addition of tsup. I should be able to do a detailed review in the next few days and get this pushed soon! 💯 

Do you know if a similar approach could be used to support Svelte?

zcf0508 committed:
feat(test): add test for .vue files

mylesmmurphy commented on packages/vscode-extension/src/hover-provider.ts:
```diff
@@ -87,4 +87,5 @@ export function registerHoverProvider(context: vscode.ExtensionContext): void {
 
```
```
hoverText.appendCodeblock(declaration + prettyTypeString, "typescript");
```
Looks like VSCode hover code blocks don't support the vue language ID, but I think its probably safe to hardcode it as "typescript" here since it's always going to be displaying a TypeScript code snippet.

zcf0508 replied:
You're right. I hardcode it to `typescript` and now the hover text is highlighted in vue files too.

mylesmmurphy commented on .gitignore:
```diff
@@ -23,4 +23,4 @@ packages/*/out
 
 # Test Artifacts
-test/workspace/.vscode
+# test/workspace/.vscode
```
I think this can just be removed entirely :) 
Otherwise, could do:
```suggestion
test/workspace/.vscode
!test/workspace/.vscode/settings.json
```

mylesmmurphy commented on test/suite/vue.test.ts:
```diff
@@ -0,0 +1,17 @@
+
+    await openDocument("test.vue");
+
+    this.timeout(90000); // Generous timeout for server readiness
```
This unfortunately won't work in CICD pipelines (which modify how timeouts work), which is why I had to write the hacky ensureTypeScriptServerReady function. Could write an alternative version named ensureVueAndTypeScriptServerReady that does a canary check against a .vue file and waits for an expected hover to be present.

zcf0508 replied:
I redefined the `ensureTypeScriptServerReady` function to support to open a vue file and now the timeout is not needed.

mylesmmurphy commented on packages/typescript-plugin/src/type-tree/index.ts:
```diff
@@ -29,12 +30,28 @@ export function getTypeInfoAtPosition(
     checker = typeChecker;
     options = prettifyOptions;
 
+    const _position = getPositionOfLineAndCharacterForVue(
```
This is a great approach, really happy with how you implemented this and help well it worked with the existing codebase out of the box. 

My main feedback is around where the responsibility of this should lie. I think it might make more sense to call this function in the main index.ts for the plugin instead, along with a type guard there as well, something like this:

```typescript
type VueProgram = ts.Program & {
  // https://github.com/vuejs/language-tools/blob/v2.0.16/packages/typescript-plugin/index.ts#L75
  __vue__: { language: Language }; // I think you can mark this as required if we're narrowed down to this type
};

...

export function isVueProgram(program: ts.Program): program is VueProgram {
  // Could also potentially add to the vue project map here
  return "__vue__" in program;
}

export function getPositionForVue(
  ctx: {
    projectName: string;
    program: VueProgram; // function would expect a VueProgram type
    typeChecker: ts.TypeChecker;
    ts: typeof ts;
    sourceFile: ts.SourceFile;
  },
  fileName: string,
  startPos = -1,
): number {
```

and in index.ts, something like this:
```typescript
      if (isVueProgram(program)) {
        position = getPositionForVue(...);
      }
```

zcf0508 replied:
Your description is very insightful!

mylesmmurphy reviewed (changes requested):
This is really great, have some minor feedback but really thankful you solved for this! 🙏 Ideally should also update the extension readme with notes indicating Vue is supported and the steps required to enable it (hybrid mode).

If you don't have time to implement this feedback, I'd be happy to do so myself, thanks again!! 😄 

P.s. if you'd like, feel free to add your name to the list of contributors in the CONTRIBUTORS.md

zcf0508 committed:
feat: always use `typescript` to render type strings

zcf0508 committed:
feat(debug): better debug experience

zcf0508 committed:
chore: update gitignore and test vscode settings

- Ignore test/workspace/.vscode directory but keep settings.json
- Emphasize importance of hybrid mode test configuration

zcf0508 committed:
refactor: add `isVueProgram` type guard for better type safety

zcf0508 committed:
fix: test failure in github ci

zcf0508 commented on packages/typescript-plugin/tsup.config.ts:
```diff
@@ -0,0 +1,12 @@
+  entry: ["src/index.ts"],
+  outDir: "out",
+  splitting: false,
+  sourcemap: false,
```
I add a new script `"build:debug": "tsup --sourcemap"` to generate sourcemap files for debugging.

zcf0508 committed:
docs: add prerequisites for Vue projects

zcf0508 committed:
chore: update contributors

zcf0508 commented:
Thanks for your reviews. I added the prerequisites for Vue projects.

zcf0508 committed:
docs: fix Vue prerequisites section in README

zcf0508 committed:
fix(vue): make __vue__ property required in VuePrograme type

mylesmmurphy reviewed (approved):
Great work on this - thanks for the thorough implementation! 👍 I’m approving it now and will make sure it gets into the next release!

One quick question before final merge ~ I am curious on the need for the `vueProgrameMap` if we're passing the program every time? 🤔

zcf0508 committed:
refactor: remove unnecessary programe cache map

mylesmmurphy commented:
Hey @zcf0508! I’m so sorry for the delay, something came up over the last few days I’ve had to attend to. This PR is looking really solid and I plan to merge it in the next few days, with a release to follow soon after! Thank you so much again for your help with this one.

mylesmmurphy reviewed (approved):
Great changes, thank you again and again sorry for the delay. This will be included in the next release pending completion of #83 😄

zcf0508 commented:
[prettify-ts-0.3.9-with-vue.zip](https://github.com/user-attachments/files/21442049/prettify-ts-0.3.9-with-vue.zip)
